File: android/src/com/dontflymoney/view/ExtractActivity.java
package com.dontflymoney.view;

import java.lang.reflect.Field;
import java.text.SimpleDateFormat;
import java.util.Calendar;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.annotation.SuppressLint;
import android.app.DatePickerDialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.DatePicker;
import android.widget.TableLayout;
import android.widget.TableRow;

import com.dontflymoney.api.Request;
import com.dontflymoney.api.Step;

public class ExtractActivity extends SmartActivity
{
	TableLayout main;
	String accounturl;

	DatePickerDialog dialog;
	private int month;
	private int year;
	
	public ExtractActivity()
	{
		init(this, R.layout.activity_extract, R.menu.extract);
	}
	
	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setCurrentInfo();
		setDate();
		getExtract();
	}
	
	private void setCurrentInfo()
	{
		main = (TableLayout)findViewById(R.id.main_table);
		accounturl = getIntent().getStringExtra("accounturl");
	}
	
	private void setDate()
	{
		Calendar today = Calendar.getInstance();
		int startMonth = getIntent().getIntExtra("month", today.get(Calendar.MONTH));
		int startYear = getIntent().getIntExtra("year", today.get(Calendar.YEAR));
		setDate(startMonth, startYear);
	}

	@SuppressLint("SimpleDateFormat")
	private void setDate(int month, int year)
	{
		this.month = month;
		this.year = year;
		
		Calendar date = Calendar.getInstance();
	    date.set(Calendar.MONTH, month);
	    date.set(Calendar.YEAR, year);
	    
	    SimpleDateFormat formatter = new SimpleDateFormat("MMM/yyyy");
	    String dateInFull = formatter.format(date.getTime());
		
	    setValue(R.id.reportDate, dateInFull);
	}
	
    public void changeDate(View v)
    {
		if(dialog == null)
		{
			dialog = new DatePickerDialog(this, new PickDate(), year, month, 1);

			try
			{
		        Field pickerField = dialog.getClass().getDeclaredField("mDatePicker");
		        pickerField.setAccessible(true);
		        DatePicker datePicker = (DatePicker) pickerField.get(dialog);
	            
	            Field field = datePicker.getClass().getDeclaredField("mDaySpinner");
	            field.setAccessible(true);
	            Object dayPicker = field.get(datePicker);
	            ((View) dayPicker).setVisibility(View.GONE);
		    } 
		    catch (Exception e) { }
		}
		
		dialog.show();
    }
	
	private class PickDate implements DatePickerDialog.OnDateSetListener
	{
	    @Override
	    public void onDateSet(DatePicker view, int year, int month, int day)
	    {
	        setDate(month, year); 
			getExtract();
	        dialog.hide();
	    }
	    
	}
	
	private void getExtract()
	{
		main.removeAllViews();
		
		Request request = new Request(this, "Moves/Extract");
		
		request.AddParameter("ticket", Authentication.Get());
		request.AddParameter("accounturl", accounturl);
		request.AddParameter("id", year * 100 + month + 1);
		
		request.Post();
	}
	
	@Override
	protected void HandleSuccess(JSONObject data, Step step) throws JSONException
	{
		JSONArray moveList = data.getJSONArray("MoveList");
		
		if (moveList.length() == 0)
		{
			View empty = createText(getString(R.string.no_extract), Gravity.CENTER);
			main.addView(empty);
		}
		else
		{
			for(int a = 0; a < moveList.length(); a++)
			{
				int color = a % 2 == 0 ? Color.TRANSPARENT : Color.LTGRAY;
				
				getMove(moveList.getJSONObject(a), color);
			}
		}
	}

	private void getMove(JSONObject move, int color)
			throws JSONException
	{
		TableRow row = new TableRow(this);
		row.setBackgroundColor(color);
		
		String description = move.getString("Description");
		row.addView(createText(description, Gravity.LEFT));

		double total = move.getDouble("Total");
		row.addView(createText(total, Gravity.RIGHT));
		
		setClick(row);

		main.addView(row);
	}
	
	private void setClick(TableRow row)
	{
		row.setClickable(true);
		
		row.setOnClickListener(new OnClickListener()
		{
		    public void onClick(View row)
		    {
		        //TableRow tablerow = (TableRow)row;
		        //TextView sample = (TextView) tablerow.getChildAt(0);
		        //alertError(sample.getText().toString());
		        
		        //redirect(MovesActivity.class);
		    }
		});
	}
	
	
	public void refresh(MenuItem menuItem)
	{
		getExtract();
	}
	
	public void goToSummary(MenuItem item)
	{
		Intent intent = new Intent(this, SummaryActivity.class);
		
		intent.putExtra("accounturl", accounturl);
		intent.putExtra("year", year);
		
		startActivity(intent);
	}
	
	public void goToMove(MenuItem item)
	{
		Intent intent = new Intent(this, MoveActivity.class);
		
		intent.putExtra("accounturl", accounturl);
		intent.putExtra("year", year);
		intent.putExtra("month", month);
		
		startActivity(intent);
	}


}

File: android/DFM/src/main/kotlin/com/darakeon/dfm/activities/ExtractActivity.kt
package com.darakeon.dfm.activities

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.view.ContextMenu
import android.view.MenuItem
import android.view.View
import android.widget.ListView
import android.widget.TextView
import com.darakeon.dfm.R
import com.darakeon.dfm.activities.base.*
import com.darakeon.dfm.activities.objects.ExtractStatic
import com.darakeon.dfm.api.InternalRequest
import com.darakeon.dfm.uiHelpers.adapters.MoveAdapter
import com.darakeon.dfm.uiHelpers.views.MoveLine
import com.darakeon.dfm.user.getAuth
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.*

class ExtractActivity : SmartActivity<ExtractStatic>(ExtractStatic), IYesNoDialogAnswer {
	private val main: ListView get() = findViewById(R.id.main_table)
	private val empty: TextView get() = findViewById(R.id.empty_list)

	private val accountUrl: String get() = getExtraOrUrl("accountUrl")

	private val dialog: DatePickerDialog
		get() = getDateDialog(
			{ y, m, _ -> updateScreen(y, m) },
			static.year, static.month
		)

	override fun contentView(): Int = R.layout.extract
	override fun contextMenuResource(): Int = R.menu.move_options
	override fun viewWithContext(): Int = R.id.main_table

	private fun updateScreen(year: Int, month: Int) {
		setDate(month, year)
		getExtract()
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		if (rotated && succeeded) {
			setDateFromLast()
			fillMoves()
		} else {
			setDateFromCaller()
			getExtract()
		}
	}

	private fun setDateFromLast() {
		setDate(static.month, static.year)
	}

	private fun setDateFromCaller() {
		if (query.containsKey("id")) {
			val id = query["id"]?.toInt() ?: 0
			val startMonth = id.mod(100) - 1
			val startYear = id / 100
			setDate(startMonth, startYear)
		} else {
			val today = Calendar.getInstance()
			val startMonth = intent.getIntExtra("month", today.get(Calendar.MONTH))
			val startYear = intent.getIntExtra("year", today.get(Calendar.YEAR))
			setDate(startMonth, startYear)
		}
	}

	private fun setDate(month: Int, year: Int) {
		static.month = month
		static.year = year

		intent.putExtra("month", month)
		intent.putExtra("year", year)

		val date = Calendar.getInstance()
		date.set(Calendar.MONTH, month)
		date.set(Calendar.YEAR, year)

		val formatter = SimpleDateFormat("MMM/yyyy", Locale.getDefault())
		val dateInFull = formatter.format(date.time)

		setValue(R.id.reportChange, dateInFull)
	}

	fun changeDate(@Suppress(ON_CLICK) view: View) {
		dialog.show()
	}

	private fun getExtract() {
		val request = InternalRequest(
			this, "Moves/Extract", { d -> handleMoves(d) }
		)

		request.addParameter("ticket", getAuth())
		request.addParameter("accountUrl", accountUrl)
		request.addParameter("id", static.year * 100 + static.month + 1)

		request.post()
	}

	private fun handleMoves(data: JSONObject) {
		static.moveList = data.getJSONArray("MoveList")
		static.name = data.getString("Name")
		static.total = data.getDouble("Total")
		static.canCheck = data.getBoolean("CanCheck")

		fillMoves()
	}

	private fun fillMoves() {
		setValue(R.id.totalTitle, static.name)
		setValueColored(R.id.totalValue, static.total)

		if (static.moveList.length() == 0) {
			main.visibility = View.GONE
			empty.visibility = View.VISIBLE
		} else {
			main.visibility = View.VISIBLE
			empty.visibility = View.GONE

			val accountAdapter = MoveAdapter(this, static.moveList, static.canCheck)
			main.adapter = accountAdapter
		}
	}


	fun goToSummary(@Suppress(ON_CLICK) view: View) {
		val intent = Intent(this, SummaryActivity::class.java)

		intent.putExtra("accountUrl", accountUrl)
		intent.putExtra("year", static.year)

		startActivity(intent)
	}

	private fun goToMove(moveId: Int) {

		val extras = Bundle()

		extras.putInt("id", moveId)
		extras.putString("accountUrl", accountUrl)
		extras.putInt("year", static.year)
		extras.putInt("month", static.month)

		createMove(extras)
	}


	override fun onContextItemSelected(item: MenuItem): Boolean {
		when (item.itemId) {
			R.id.edit_move -> {
				edit()
				return true
			}
			R.id.delete_move -> {
				askDelete()
				return true
			}
			R.id.check_move -> {
				check()
				return true
			}
			R.id.uncheck_move -> {
				uncheck()
				return true
			}
			else -> return super.onContextItemSelected(item)
		}
	}

	private fun edit() {
		val moveLine = clickedView as MoveLine
		goToMove(moveLine.id)
	}


	private fun askDelete(): Boolean {
		var messageText = getString(R.string.sure_to_delete)
		val moveRow = clickedView as MoveLine

		messageText = String.format(messageText, moveRow.description)

		alertYesNo(messageText, this)

		return false
	}

	override fun yesAction() {
		submitMoveAction("Delete")
	}

	override fun noAction() {}


	private fun check() {
		submitMoveAction("Check")
	}

	private fun uncheck() {
		submitMoveAction("Uncheck")
	}


	private fun submitMoveAction(action: String) {
		val view = clickedView as MoveLine

		val request = InternalRequest(
			this, "Moves/" + action, { refresh() }
		)

		request.addParameter("ticket", getAuth())
		request.addParameter("accountUrl", accountUrl)
		request.addParameter("id", view.id)

		request.post()
	}


	public override fun changeContextMenu(view: View, menuInfo: ContextMenu) {
		val move = clickedView as MoveLine

		try {
			if (static.canCheck) {
				hideMenuItem(menuInfo, if (move.isChecked) R.id.check_move else R.id.uncheck_move)
				showMenuItem(menuInfo, if (move.isChecked) R.id.uncheck_move else R.id.check_move)
			} else {
				hideMenuItem(menuInfo, R.id.check_move)
				hideMenuItem(menuInfo, R.id.uncheck_move)
			}
		} catch (e: Exception) {
			e.printStackTrace()
		}

	}

	private fun hideMenuItem(menu: ContextMenu, id: Int) {
		toggleMenuItem(menu, id, false)
	}

	private fun showMenuItem(menu: ContextMenu, id: Int) {
		toggleMenuItem(menu, id, true)
	}

	private fun toggleMenuItem(menu: ContextMenu, id: Int, show: Boolean) {
		val buttonToHide = menu.findItem(id)
		buttonToHide.isVisible = show
	}

}

